projects/plugins/jetpack/_inc/lib/class.media-extractor.php

```diff
@@ -399,29 +401,67 @@ private static function get_image_fields( $post, $args = array() ) {
 
 		$from_featured_image = Jetpack_PostImages::from_thumbnail( $post->ID, $args['width'], $args['height'] );
 		if ( ! empty( $from_featured_image ) ) {
-			$srcs       = wp_list_pluck( $from_featured_image, 'src' );
-			$image_list = array_merge( $image_list, $srcs );
+			if ( $extract_alt_text ) {
+				$image_list = array_merge( $image_list, self::reduce_extracted_images( $from_featured_image ) );
+			} else {
+				$srcs       = wp_list_pluck( $from_featured_image, 'src' );
+				$image_list = array_merge( $image_list, $srcs );
+			}
 		}
 
 		$from_slideshow = Jetpack_PostImages::from_slideshow( $post->ID, $args['width'], $args['height'] );
 		if ( ! empty( $from_slideshow ) ) {
-			$srcs       = wp_list_pluck( $from_slideshow, 'src' );
-			$image_list = array_merge( $image_list, $srcs );
+			if ( $extract_alt_text ) {
+				$image_list = array_merge( $image_list, self::reduce_extracted_images( $from_slideshow ) );
+			} else {
+				$srcs       = wp_list_pluck( $from_slideshow, 'src' );
+				$image_list = array_merge( $image_list, $srcs );
+			}
 		}
 
 		$from_gallery = Jetpack_PostImages::from_gallery( $post->ID );
 		if ( ! empty( $from_gallery ) ) {
-			$srcs       = wp_list_pluck( $from_gallery, 'src' );
-			$image_list = array_merge( $image_list, $srcs );
+			if ( $extract_alt_text ) {
+				$image_list = array_merge( $image_list, self::reduce_extracted_images( $from_gallery ) );
+			} else {
+				$srcs       = wp_list_pluck( $from_gallery, 'src' );
+				$image_list = array_merge( $image_list, $srcs );
+			}
 			$image_booleans['gallery']++; // @todo This count isn't correct, will only every count 1
 		}
 
 		// @todo Can we check width/height of these efficiently?  Could maybe use query args at least, before we strip them out
-		$image_list = self::get_images_from_html( $post->post_content, $image_list );
+		$image_list = self::get_images_from_html( $post->post_content, $image_list, $extract_alt_text );
 
 		return self::build_image_struct( $image_list, $image_booleans );
 	}
 
+	/**
+	 * Given an extracted image array reduce to src and alt_text.
+	 *
+	 * @param array $images extracted image array.
+	 *
+	 * @return array reduced image array
+	 */
+	protected static function reduce_extracted_images( $images ) {
+		$ret_images = array();
+		foreach ( $images as $image ) {
+			// skip if src isn't set.
+			if ( empty( $image['src'] ) ) {
+				continue;
+			}
+			if ( ! empty( $image['alt_text'] ) ) {
+				$ret_images[] = array(
+					'url'      => $image['src'],
+					'alt_text' => $image['alt_text'],
+				);
+			} else {
+				$ret_images[] = $image['src'];
+			}
```

Comment:
What if our array always had the same structure, with something like:

```suggestion
			$ret_images[] = array(
				'url'      => $image['src'],
				'alt_text' => ! empty( $image['alt_text'] ) ? $image['alt_text'] : '',
			);
```

That may allow us to avoid special handling later, like in line 491?

Reply:
We'd still need special handling as for this to be backwards compatible the alt_text can't be set on the returned data. If it is present it will lead to an indexing issues.

The diff is structured to allow us to limit this update to only the code flow used by Jetpack search while remaining the integrity of the flow used for other indexing purposes.